nni/retiarii/hub/pytorch/autoformer.py

```diff
@@ -0,0 +1,464 @@
+        # get the embeddings with the corresponding distance
+        embeddings = self.embeddings_table_v[final_mat_v] + \
+            self.embeddings_table_h[final_mat_h]
+
+        return embeddings
+
+
+class Attention(nn.Module):
+    def __init__(
+            self,
+            embed_dim,
+            fixed_embed_dim,
+            num_heads,
+            attn_drop=0.,
+            proj_drop=0,
+            rpe=False,
+            qkv_bias=False,
+            qk_scale=None,
+            rpe_length=14) -> None:
+        super().__init__()
+        self.num_heads = num_heads
+        head_dim = embed_dim // num_heads
+        self.scale = qk_scale or head_dim ** -0.5
+        self.qkv = nn.Linear(embed_dim, embed_dim * 3, bias=qkv_bias)
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.proj = nn.Linear(embed_dim, embed_dim)
+        self.proj_drop = nn.Dropout(proj_drop)
+        self.rpe = rpe
+        if rpe:
+            self.rel_pos_embed_k = RelativePosition2D(
+                fixed_embed_dim // num_heads, rpe_length)
+            self.rel_pos_embed_v = RelativePosition2D(
+                fixed_embed_dim // num_heads, rpe_length)
+
+    def forward(self, x):
+        B, N, C = x.shape
+        qkv = self.qkv(x).reshape(
+            B,
+            N,
+            3,
+            self.num_heads,
+            C //
+            self.num_heads).permute(
+            2,
+            0,
+            3,
+            1,
+            4)
+        # make torchscript happy (cannot use tensor as tuple)
+        q, k, v = qkv[0], qkv[1], qkv[2]
+
+        attn = (q @ k.transpose(-2, -1)) * self.scale
+        if self.rpe:
+            r_p_k = self.rel_pos_embed_k(N, N)
+            attn = attn + (
+                q.permute(
+                    2, 0, 1, 3).reshape(
+                    N, self.num_heads * B, -1) @ r_p_k.transpose(
+                    2, 1)) .transpose(
+                1, 0).reshape(
+                        B, self.num_heads, N, N) * self.scale
+
+        attn = attn.softmax(dim=-1)
+        attn = self.attn_drop(attn)
+
+        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
+        if self.rpe:
+            r_p_v = self.rel_pos_embed_v(N, N)
+            attn_1 = attn.permute(
+                2, 0, 1, 3).reshape(
+                N, B * self.num_heads, -1)
+            # The size of attention is (B, num_heads, N, N), reshape it to (N, B*num_heads, N) and do batch matmul with
+            # the relative position embedding of V (N, N, head_dim) get shape like (N, B*num_heads, head_dim). We reshape it to the
+            # same size as x (B, num_heads, N, hidden_dim)
+            x = x + (attn_1 @ r_p_v).transpose(1, 0).reshape(B,
+                                                             self.num_heads, N, -1).transpose(2, 1).reshape(B, N, -1)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+
+class TransformerEncoderLayer(nn.Module):
```

Comment:
Suggest adding docstrings here to explain what's the difference between this encoder layer and pytorch's built-in encoder layer.